### ISSUE TITLE
Good combination of all the perfect attributes for a great application running

### ISSUE THREAD

**Issue by hatewesterncivilization** [BiglySoftware/BiglyBT#121]
Hello all! Looks like I've found finally an great combination of at one blow BiglyBT's stability, pure UI responsibility, big throughput, low pauses, speed and so on. Also memory metric\usage showing very good result.

Anyone please test this thing. Just do not forget using Java 1.8 instead of 1.9, because 1.9 start discontinue support of Concurrent Garbage Collector (-XX:+UseConcMarkSweepGC), what's we actually need for the low pauses. Also use 64-bit versions of Java & SWT.    

-Dazureus.config.path=G:\Azureusc2\\
#-server
-Xgc:throughput
#-XpauseTarget=150
-Xrs
#-XgcPrio:deterministic 
-d64
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
-Xquickstart
#-Xincgc
-Xnoclassgc
-Xverify:all
#-XX:+ForceLargePagesForHeap
-XX:OptThreads=8
#-XlargePages:exitOnFailure=true
#-XX:+TieredCompilation
#-XX:CompileThreshold=1000
#-XX:-DontCompileHugeMethods 
#-XX:CICompilerCount=1
-XX:ReservedCodeCacheSize=1g
#-XX:MaxInlineSize=10240  
#-XX:FreqInlineSize=10240
#-XX:+AggressiveOpts
-XX:+AlwaysPreTouch
-XX:UseSSE=3
#-XX:+UseLargePages
#-XX:LargePageSizeInBytes=2m
#-XX:-BytecodeVerificationLocal 
#-XX:-BytecodeVerificationRemote 
-XX:+PerfDisableSharedMem
-XX:SoftRefLRUPolicyMSPerMB=1
#-XX:MaxMetaspaceSize=250m
#-Xss60m
-Xms1650m
-Xmx1650m
-Xmn150m
-XX:MaxDirectMemorySize=800m
#-XX:+UseThreadPriorities
#-XX:ThreadPriorityPolicy=1
-XX:+UseCompressedOops
-XX:+DoEscapeAnalysis
#-XX:+EliminateLocks
-XX:+UseFastAccessorMethods
-XX:+UseTLAB
-XX:+DisableExplicitGC
-XX:+UseParNewGC
-XX:+CMSParallelRemarkEnabled
-XX:+UseCMSInitiatingOccupancyOnly 
#-XX:+ScavengeBeforeFullGC 
-XX:+UseConcMarkSweepGC 
#-XX:+CMSScavengeBeforeRemark
#-XX:+CMSConcurrentMTEnabled
-XX:ParallelGCThreads=8
-XX:+ExplicitGCInvokesConcurrent 
#-XX:NewRatio=2
#-XX:GCTimeRatio=2 
-XX:MaxGCPauseMillis=100
#-XX:PermSize=128m
#-XX:MaxPermSize=128m
#-XX:NewSize=128m
#-XX:MaxNewSize=128m
-XX:+CMSClassUnloadingEnabled
-XX:SurvivorRatio=8
-XX:TargetSurvivorRatio=95
#-XX:+CMSIncrementalMode 
#-XX:+CMSIncrementalPacing 
#-XX:CMSIncrementalDutyCycleMin=0 
#-XX:CMSIncrementalDutyCycle=10 
-XX:MaxTenuringThreshold=0 
-XX:CMSInitiatingOccupancyFraction=80
#-XX:+UseAdaptiveGCBoundary
-XX:-UseGCOverheadLimit

**Comment by the8472** [comment]
Unless you have a justification or measurements for every parameter there this looks like cargo-culting.

**Comment by hatewesterncivilization** [comment]
the8472, thanks man for a good review, it's a big honor for me, I know, you are one of the first developers, once again thanks. Of course, you are clearly right, concerning my way, I'm evaluate improvements in performances just visually and approximately, yes, subjectively, but this method is really work.

Here is some modifications of above, seems some improvements

-Dazureus.config.path=G:\Azureusc2\\
#-server
-Xgc:throughput
#-XpauseTarget=150
-Xrs
#-XgcPrio:deterministic 
-d64
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
-Xquickstart
#-Xincgc
-Xnoclassgc
-Xverify:all
#-XX:+ForceLargePagesForHeap
-XX:OptThreads=8
#-XlargePages:exitOnFailure=true
#-XX:+TieredCompilation
#-XX:CompileThreshold=1000
#-XX:-DontCompileHugeMethods 
#-XX:CICompilerCount=1
-XX:ReservedCodeCacheSize=1g
#-XX:MaxInlineSize=10240  
#-XX:FreqInlineSize=10240
#-XX:+AggressiveOpts
-XX:+AlwaysPreTouch
-XX:UseSSE=3
#-XX:+UseLargePages
#-XX:LargePageSizeInBytes=2m
#-XX:-BytecodeVerificationLocal 
#-XX:-BytecodeVerificationRemote 
-XX:+PerfDisableSharedMem
-XX:SoftRefLRUPolicyMSPerMB=1
-XX:MaxMetaspaceSize=250m
-Xss60m
-Xms1650m
-Xmx1650m
-Xmn150m
-XX:MaxDirectMemorySize=1650m
#-XX:+UseThreadPriorities
#-XX:ThreadPriorityPolicy=1
-XX:+UseCompressedOops
-XX:+DoEscapeAnalysis
#-XX:+EliminateLocks
-XX:+UseFastAccessorMethods
-XX:+UseTLAB
-XX:+DisableExplicitGC
-XX:+UseParNewGC
-XX:+CMSParallelRemarkEnabled
-XX:+UseCMSInitiatingOccupancyOnly 
#-XX:+ScavengeBeforeFullGC 
-XX:+UseConcMarkSweepGC 
#-XX:+CMSScavengeBeforeRemark
#-XX:+CMSConcurrentMTEnabled
-XX:ParallelGCThreads=8
-XX:+ExplicitGCInvokesConcurrent 
#-XX:NewRatio=2
#-XX:GCTimeRatio=2 
-XX:MaxGCPauseMillis=100
#-XX:PermSize=128m
#-XX:MaxPermSize=128m
#-XX:NewSize=128m
#-XX:MaxNewSize=128m
-XX:+CMSClassUnloadingEnabled
-XX:SurvivorRatio=16
-XX:TargetSurvivorRatio=95
#-XX:+CMSIncrementalMode 
#-XX:+CMSIncrementalPacing 
#-XX:CMSIncrementalDutyCycleMin=0 
#-XX:CMSIncrementalDutyCycle=10 
-XX:MaxTenuringThreshold=0 
-XX:CMSInitiatingOccupancyFraction=80
#-XX:+UseAdaptiveGCBoundary
-XX:-UseGCOverheadLimit

**Comment by the8472** [comment]
First of all, posting all the commented out parameters only adds visual noise.

Now to the options you're actually setting:

>-Dazureus.config.path=G:\Azureusc2\

this is specific to your PC

>-Xgc:throughput

This is a jrockit JVM option, I don't think it applies to the most commonly installed JVM (oracle/hotspot)

>-Xrs

reduced signal usage generally would have a negative impact on performance since it disables some optimizations

>-d64

does nothing these days

>-Xquickstart

This is an IBM JVM option

>-Xnoclassgc

might have a slightly negative impact if plugins get unloaded/reloaded during updates

>-Xverify:all

this is a security thing, has nothing to do with performance

>-XX:OptThreads=8

again, a jrockit option

>-XX:+AlwaysPreTouch

doesn't make much of a difference except on very large heaps

>-XX:UseSSE=3

this is auto-detected in a machine-specific manner, no tuning needed

>-XX:+PerfDisableSharedMem

this is only needed if you have a slow drive on which the temp directory is located

>-Xss60m

That's some huge stack, generally not necessary

>-Xms1650m
>-Xmx1650m
>-Xmn150m
>-XX:MaxDirectMemorySize=1650m

specific to your workload and your tuning goals, not really useful as general advise


>-XX:+UseCompressedOops

already the default in most cases

>-XX:+DoEscapeAnalysis

already the default

>-XX:+UseFastAccessorMethods

enabling them [worsens performance](https://bugs.openjdk.java.net/browse/JDK-8013392)

>-XX:+UseTLAB

already the default

>-XX:+DisableExplicitGC

it would be considerably better to use `+ExplicitGCInvokesConcurrent` since Bigly may rely on those to avoid OOMs.

>-XX:+UseParNewGC

This is already the default if you enable CMS

>-XX:+CMSParallelRemarkEnabled

already the default

>-XX:+UseCMSInitiatingOccupancyOnly

generally a good idea with CMS, but only if you also tune the IHOP based on GC logs

>-XX:+UseConcMarkSweepGC

CMS is nice for responsiveness, yes.

>-XX:ParallelGCThreads=8

this is generally auto-tuned based on available hardware and should not be considered general advise

>-XX:+ExplicitGCInvokesConcurrent

well, that is useless since you disabled explicit above

>-XX:TargetSurvivorRatio=95

this would only be advisable after observing GC logs, otherwise it can lead to promotion bursts

>-XX:MaxTenuringThreshold=0

very bad idea, leads to immediate object promotion

>-XX:CMSInitiatingOccupancyFraction=80

ah, so you're setting the IHOP after all. curious that it is not grouped with the related option.
anyway, this usually only makes sense after examining GC logs.

>-XX:-UseGCOverheadLimit

there rarely is a good reason to disable it because the exception being thrown already indicates some other problem. and it has no impact on performance, only on error behavior


----

In summary, most of your settings don't make sense, and even those that do make sense may be specific to your machine or how you use bigly. So sharing them with other people probably doesn't help themand perpetuates cargo-culting.

I would suggest removing most options and asking for advise if you're experiencing some specific, quantifiable problems.

**Comment by hatewesterncivilization** [comment]
the8472, thanks, that's what we looking for, detailed explanations about specific jvm settings.

Can I drop-share some fresh version of .vmoptions file? Now it's really stable in long running meaning and of course low pause and good UI responsiveness. Sounds like -client helps better than -server and tiered mode.

-Dazureus.config.path=G:\Azureusc2\\
-client
-d64
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
#-XX:+TieredCompilation
#-XX:OptThreads=8
#-XX:ReservedCodeCacheSize=500m
-Xms1650m
-Xmx1650m
-Xmn50m
#-XX:+AlwaysPreTouch
-XX:+PerfDisableSharedMem
-XX:+ParallelRefProcEnabled
-XX:+UseConcMarkSweepGC
-XX:+UseParNewGC
-XX:ParallelGCThreads=8
#-XX:ParallelCMSThreads=4
-XX:+CMSParallelRemarkEnabled
-XX:CMSInitiatingOccupancyFraction=92 
-XX:+UseCMSInitiatingOccupancyOnly
-XX:CMSInitiatingOccupancyFraction=70
-XX:MaxTenuringThreshold=2
-XX:+ScavengeBeforeFullGC 
-XX:+CMSScavengeBeforeRemark
#-XX:ParGCCardsPerStrideChunk=32768
#-XX:+BindGCTaskThreadsToCPUs
#-XX:+UseGCTaskAffinity
-XX:SurvivorRatio=16
-XX:+CMSClassUnloadingEnabled
-XX:MaxGCPauseMillis=100
#-XX:+UseCMSCompactAtFullCollection
-XX:+DisableExplicitGC

**Comment by hatewesterncivilization** [comment]
-Dazureus.config.path=J:\BiglyBT\
-client
-d64
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
#-XX:+TieredCompilation
#-XX:OptThreads=8
#-XX:ReservedCodeCacheSize=500m
-Xms1650m
-Xmx1650m
-Xmn50m
#-XX:+AlwaysPreTouch
-XX:+PerfDisableSharedMem
-XX:+ParallelRefProcEnabled
-XX:+UseConcMarkSweepGC
-XX:+UseParNewGC
-XX:ParallelGCThreads=8
#-XX:ParallelCMSThreads=4
-XX:+CMSParallelRemarkEnabled
-XX:+UseCMSInitiatingOccupancyOnly
-XX:CMSInitiatingOccupancyFraction=92
-XX:MaxTenuringThreshold=2
-XX:+ScavengeBeforeFullGC 
-XX:+CMSScavengeBeforeRemark
#-XX:ParGCCardsPerStrideChunk=32768
#-XX:-BindGCTaskThreadsToCPUs
#-XX:-UseGCTaskAffinity
-XX:SurvivorRatio=16
-XX:+CMSClassUnloadingEnabled
-XX:MaxGCPauseMillis=100
#-XX:+UseCMSCompactAtFullCollection
-XX:+DisableExplicitGC

**Comment by hatewesterncivilization** [comment]
-Dazureus.config.path=G:\Azureusc2\\
#-client
-Xgc:throughput
-d32
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
#-XX:+TieredCompilation
#-XX:OptThreads=8
-XX:ReservedCodeCacheSize=700m
#-Xss60m
-Xms1650m
-Xmx1650m
-Xmn250m
-XX:+AlwaysPreTouch
-XX:+PerfDisableSharedMem
-XX:+ParallelRefProcEnabled
-XX:+UseConcMarkSweepGC
-XX:+UseParNewGC
#-XX:ParallelGCThreads=16
#-XX:ParallelCMSThreads=4
-XX:+CMSParallelRemarkEnabled
-XX:+UseCMSInitiatingOccupancyOnly
-XX:CMSInitiatingOccupancyFraction=42
-XX:MaxTenuringThreshold=0
-XX:+ScavengeBeforeFullGC
-XX:+CMSScavengeBeforeRemark
#-XX:ParGCCardsPerStrideChunk=32768
-XX:-BindGCTaskThreadsToCPUs
-XX:-UseGCTaskAffinity
-XX:SurvivorRatio=16
-XX:+CMSClassUnloadingEnabled
-XX:MaxGCPauseMillis=1000
#-XX:+UseCMSCompactAtFullCollection
-XX:+DisableExplicitGC

**Comment by hatewesterncivilization** [comment]
Hi, all. This set is completely the best, -Xgc:pausetime and do not specifying of -XX:SurvivorRatio does the most thing for fantastic low pause, responsiveness as well as speed & throughput. Let's try and merry christmas Bigly, devs and all of the peoples!

-Dazureus.config.path=G:\Azureusc2\\
-Dsun.awt.keepWorkingSetOnMinimize=true
-Dsun.java2d.noddraw=true
-Dsun.java2d.d3d=false
#-server
-Xgc:pausetime
-XpauseTarget=150
-Xrs
-d32
#-Djava.awt.headless=true
-XX:+UnlockExperimentalVMOptions
-XX:+UnlockDiagnosticVMOptions
-Xquickstart
-Xnoclassgc
-Xverify:all
#-XX:+ForceLargePagesForHeap
-XX:OptThreads=8
#-XlargePages:exitOnFailure=true
#-XX:+TieredCompilation
#-XX:CompileThreshold=1000
#-XX:-DontCompileHugeMethods
#-XX:CICompilerCount=1
-XX:ReservedCodeCacheSize=400m
#-XX:MaxInlineSize=10240
#-XX:FreqInlineSize=10240
#-XX:+AggressiveOpts
#-XX:+AlwaysPreTouch
-XX:UseSSE=3
#-XX:+UseLargePages
#-XX:LargePageSizeInBytes=2m
#-XX:-BytecodeVerificationLocal
#-XX:-BytecodeVerificationRemote
-XX:+PerfDisableSharedMem
-XX:+ParallelRefProcEnabled
-XX:SoftRefLRUPolicyMSPerMB=1
#-XX:MaxMetaspaceSize=250m
#-Xss60m
-Xms1650m
-Xmx1650m
-Xmn150m
-XX:MaxDirectMemorySize=1400m
#-XX:+UseThreadPriorities
#-XX:ThreadPriorityPolicy=1
-XX:+UseCompressedOops
-XX:+DoEscapeAnalysis
#-XX:+EliminateLocks
-XX:+UseFastAccessorMethods
-XX:+UseTLAB
-XX:+DisableExplicitGC
-XX:+UseParNewGC
-XX:+UseConcMarkSweepGC
#-XX:+CMSParallelRemarkEnabled
-XX:+UseCMSInitiatingOccupancyOnly
#-XX:+ScavengeBeforeFullGC
#-XX:+CMSScavengeBeforeRemark
#-XX:+CMSConcurrentMTEnabled
#-XX:ParallelGCThreads=8
-XX:+ExplicitGCInvokesConcurrent
#-XX:NewRatio=2
#-XX:GCTimeRatio=2
-XX:MaxGCPauseMillis=150
#-XX:PermSize=128m
#-XX:MaxPermSize=128m
#-XX:NewSize=128m
#-XX:MaxNewSize=128m
-XX:+CMSClassUnloadingEnabled
#-XX:SurvivorRatio=16
-XX:TargetSurvivorRatio=90
#-XX:+CMSIncrementalMode
#-XX:+CMSIncrementalPacing
#-XX:CMSIncrementalDutyCycleMin=0
#-XX:CMSIncrementalDutyCycle=10
-XX:MaxTenuringThreshold=0
-XX:CMSInitiatingOccupancyFraction=90
#-XX:+UseAdaptiveGCBoundary
-XX:-UseGCOverheadLimit
